odoo/modules/module.py

```diff
@@ -311,74 +351,85 @@ def get_module_root(path):
         path = new_path
     return path
 
-def load_information_from_description_file(module, mod_path=None):
+@tools.func.mockable
```

Comment:
I don't think test code should be invading production code. Are we going to need to put this decorator and overhead everywhere? That's extra bloat we don't need in production.

Reply:
I'm not a fan of this solution either *but* it is the best way beside forbidding people from from-importing the function (see the `get_request_website` function of `website/models/ir_http.py`). The scope of this decorator is very limited though, it is basically useful *only* for functions and classes declared in the global scope.

I agree with you that we should not have test-dedicated stuff in production code but in this very precise situation it is reasonable (it is a limitation of the language), the solution is cheap (only extra function call) and it only requires a minimal setup (one decorator).

Reply:
Some test depends on patching the function, see odoo/addons/test_assetsbundle/tests/test_assetsbundle.py.
The point of Julien is that the former "simple patch" forbids developers to import the function as
```python
from odoo.modules import load_manifest
```
because this occurrence of `load_manifest` wouldn't be patched.

Reply:
@odony's point stands though: this is an issue which occurs with *every free function*, are we going to decorate every free function in case somebody somewhere needs to patch it? Especially here, it's being decorated for the benefit of... two tests?

The normal solution is to update the import to allow patching. It *could* make sense to update the function to allow patching it regardless if it's something that's imported everywhere as a free function *and* is being patched by lots of tests.

That is definitely not the case for `load_manifest`.

Reply:
xdo suggested me this:

    def _load_manifest(...):
        ...
        with open(...) as fd:
            manifest = ast.literal_eval(fd.read())
        ...

    @lru_cache
    def load_manifest(...):
        return _load_manifest(...)

Reply:
> EDIT: but still, wouldn't work with test_assetsbundles... Seriously just let me merge this `mockable` decorator...

No, it's shit.